server/src/main/java/org/elasticsearch/cluster/block/ClusterBlocks.java

```diff
@@ -68,27 +67,37 @@ public Map<String, Set<ClusterBlock>> indices(ClusterBlockLevel level) {
     }
 
     private Set<ClusterBlock> blocksForIndex(ClusterBlockLevel level, String index) {
-        return indices(level).getOrDefault(index, emptySet());
+        return indices(level).getOrDefault(index, Set.of());
     }
 
     private static EnumMap<ClusterBlockLevel, ImmutableLevelHolder> generateLevelHolders(
         Set<ClusterBlock> global,
         Map<String, Set<ClusterBlock>> indicesBlocks
     ) {
         EnumMap<ClusterBlockLevel, ImmutableLevelHolder> levelHolders = new EnumMap<>(ClusterBlockLevel.class);
+        List<ClusterBlock> setBuilder = new ArrayList<>();
+        Map<String, Set<ClusterBlock>> indicesBuilder = Maps.newMapWithExpectedSize(indicesBlocks.size());
         for (final ClusterBlockLevel level : ClusterBlockLevel.values()) {
-            Predicate<ClusterBlock> containsLevel = block -> block.contains(level);
-            Set<ClusterBlock> newGlobal = global.stream().filter(containsLevel).collect(toUnmodifiableSet());
-
-            Map<String, Set<ClusterBlock>> indicesBuilder = Maps.newMapWithExpectedSize(indicesBlocks.size());
             for (Map.Entry<String, Set<ClusterBlock>> entry : indicesBlocks.entrySet()) {
-                indicesBuilder.put(entry.getKey(), entry.getValue().stream().filter(containsLevel).collect(toUnmodifiableSet()));
+                indicesBuilder.put(entry.getKey(), addBlocksAtLevel(entry.getValue(), setBuilder, level));
             }
-            levelHolders.put(level, new ImmutableLevelHolder(newGlobal, Map.copyOf(indicesBuilder)));
+            levelHolders.put(level, new ImmutableLevelHolder(addBlocksAtLevel(global, setBuilder, level), Map.copyOf(indicesBuilder)));
+            indicesBuilder.clear();
         }
         return levelHolders;
     }
 
+    private static Set<ClusterBlock> addBlocksAtLevel(Set<ClusterBlock> blocks, List<ClusterBlock> setBuilder, ClusterBlockLevel level) {
+        for (ClusterBlock clusterBlock : blocks) {
+            if (clusterBlock.contains(level)) {
+                setBuilder.add(clusterBlock);
+            }
+        }
+        var res = Set.of(setBuilder.toArray(EMPTY_BLOCKS_ARRAY));
```

Comment:
Why not `setBuilder.toArray(ClusterBlock[]::new)`?

Reply:
It's a little slower I would imagine and in the end the array supplier function is just called with a `0` by the implementation anyway so why not do the JITs work ourselves here I guess :)